agentops/llm_tracker.py

```diff
@@ -621,6 +624,130 @@ async def async_generator():
+                        self.tool_id = chunk.content_block.id
+                        self.tool_event[self.tool_id] = ToolEvent(
+                            name=chunk.content_block.name,
+                            logs={"type": chunk.content_block.type, "input": ""},
+                        )
+                elif chunk.type == "content_block_delta":
+                    if chunk.delta.type == "text_delta":
+                        self.llm_event.completion["content"] += chunk.delta.text
+                    elif chunk.delta.type == "input_json_delta":
+                        self.tool_event[self.tool_id].logs[
+                            "input"
+                        ] += chunk.delta.partial_json
+                elif chunk.type == "content_block_stop":
+                    pass
+                elif chunk.type == "message_delta":
+                    self.llm_event.completion_tokens = chunk.usage.output_tokens
+                elif chunk.type == "message_stop":
+                    self.llm_event.end_timestamp = get_ISO_time()
+                    self._safe_record(session, self.llm_event)
+
+            except Exception as e:
+                self._safe_record(
+                    session, ErrorEvent(trigger_event=self.llm_event, exception=e)
+                )
+
+                kwargs_str = pprint.pformat(kwargs)
+                chunk = pprint.pformat(chunk)
+                logger.warning(
+                    f"Unable to parse a chunk for LLM call. Skipping upload to AgentOps\n"
+                    f"chunk:\n {chunk}\n"
+                    f"kwargs:\n {kwargs_str}\n",
+                )
+
+        # if the response is a generator, decorate the generator
+        if isinstance(response, Stream):
+
+            def generator():
+                for chunk in response:
+                    handle_stream_chunk(chunk)
+                    yield chunk
+
+            return generator()
+
+        # For asynchronous AsyncStream
+        if isinstance(response, AsyncStream):
+
+            async def async_generator():
+                async for chunk in response:
+                    handle_stream_chunk(chunk)
+                    yield chunk
+
+            return async_generator()
+
+        # For async AsyncMessages
+        if isinstance(response, AsyncMessages):
+
+            async def async_generator():
+                async for chunk in response:
+                    handle_stream_chunk(chunk)
+                    yield chunk
+
+            return async_generator()
+
+        # Handle object responses
+        try:
+            self.llm_event.returns = response.model_dump()
+            self.llm_event.agent_id = check_call_stack_for_agent_id()
+            self.llm_event.prompt = kwargs["messages"]
+            self.llm_event.prompt_tokens = response.usage.input_tokens
+            self.llm_event.completion = {
+                "role": "assistant",
+                "content": response.content,
+            }
```

Comment:
<img width="492" alt="image" src="https://github.com/user-attachments/assets/4f8b2b54-0666-4913-bd0e-29c715c1f756">

Issue where async completion non-stream doesn't get formatted as an LLMEvent correctly. Sample code:

```
import asyncio
from anthropic import AsyncAnthropic

client = AsyncAnthropic(
    # This is the default and can be omitted
    api_key=os.environ.get("ANTHROPIC_API_KEY"),
)


async def main() -> None:
    message = await client.messages.create(
        max_tokens=1024,
        messages=[
            {
                "role": "user",
                "content": "Hello, Claude",
            }
        ],
        model="claude-3-opus-20240229",
    )
    print(message.content)


await main()```

Reply:
Thanks for testing this out! I will push a fix soon.

Reply:
Fixed in the latest commit!